libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+}
+
+// a 'flag' callback has no optargs, so we need to set it ourself based on *this* command
+int cli_int_buildmode_flag_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  int rc = CLI_BUILDMODE_EXTEND;
+
+  if (argc) {
+    cli_error(cli, "Extra arguments on command line, command ignored.");
+    rc = CLI_ERROR;
+  }
+  if (cli_int_add_optarg_value(cli, command, command, 0)) {
+    cli_error(cli, "Problem setting value for optional flag %s", command);
+    rc = CLI_ERROR;
+  }
+  return rc;
+}
+
+// a 'flag' callback has no optargs, so we need to set it ourself based on *this* command
+int cli_int_buildmode_flag_multiple_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  int rc = CLI_BUILDMODE_EXTEND;
+
+  if (argc) {
+    cli_error(cli, "Extra arguments on command line, command ignored.");
+    rc = CLI_ERROR;
+  }
+  if (cli_int_add_optarg_value(cli, command, command, CLI_CMD_OPTION_MULTIPLE)) {
+    cli_error(cli, "Problem setting value for optional flag %s", command);
+    rc = CLI_ERROR;
+  }
+
+  return rc;
+}
+
+int cli_int_buildmode_cancel_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  int rc = CLI_BUILDMODE_CANCEL;
+
+  if (argc > 0) {
+    cli_error(cli, "Extra arguments on command line, command ignored.");
+    rc = CLI_ERROR;
+  }
+  return rc;
+}
+
+int cli_int_buildmode_exit_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  int rc = CLI_BUILDMODE_EXIT;
+
+  if (argc > 0) {
+    cli_error(cli, "Extra arguments on command line, command ignored.");
+    rc = CLI_ERROR;
+  }
+  return rc;
+}
+
+int cli_int_buildmode_show_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  struct cli_optarg_pair *optarg_pair;
+  if (cli && cli->buildmode) {
+    for (optarg_pair = cli->found_optargs; optarg_pair; optarg_pair = optarg_pair->next) {
+      // only show vars that are also current 'commands'
+      struct cli_command *c = cli->commands;
+      for (; c; c = c->next) {
+        if (c->command_type != CLI_BUILDMODE_COMMAND) continue;
+        if (!strcmp(c->command, optarg_pair->name)) {
+          cli_print(cli, "  %-20s = %s", optarg_pair->name, optarg_pair->value);
+          break;
+        }
+      }
+    }
+  }
+  return CLI_OK;
+}
+
+int cli_int_buildmode_unset_cback(struct cli_def *cli, const char *command, char *argv[], int argc) {
+  // iterate over our 'set' variables to see if that variable is also a 'valid' command right now
+
+  struct cli_command *c;
+
+  // is this 'optarg' to remove one of the current commands?
+  for (c = cli->commands; c; c = c->next) {
+    if (c->command_type != CLI_BUILDMODE_COMMAND) continue;
+    if (cli->privilege < c->privilege) continue;
+    if ((cli->buildmode->mode != c->mode) && (cli->buildmode->transient_mode != c->mode) && (c->mode != MODE_ANY))
+      continue;
+    if (strcmp(c->command, argv[0])) continue;
+    // Ok, go fry anything by this name
```

Comment:
fry?

Reply:
Delete - this routine is for use in build mode if you choose to unset something you've assigned.  In the case of a flag that could be set multiple times (think something like multiple 'verbose' flags to change the level of verbosity) it deletes *all* occurrences of that setting.